### ISSUE TITLE
Improve how to set levels on events

### ISSUE THREAD

**Issue by rhcarvalho** [getsentry/sentry-go#151]
Some users rely on multiple [levels](https://github.com/getsentry/sentry-go/blob/ea2e5149b29d63daf01b0cad6b3a54d8b60b08a6/interfaces.go#L16-L25) when reporting events to Sentry.

The current API makes it very inconvenient to write code that reports events with different levels in the same code block:

```go
func foo() {
    sentry.WithScope(func(scope *sentry.Scope) {
        scope.SetLevel(sentry.LevelWarning)
        sentry.CaptureException(errors.New("example warning"))
    })
    sentry.WithScope(func(scope *sentry.Scope) {
        scope.SetLevel(sentry.LevelFatal)
        sentry.CaptureException(errors.New("example error"))
    })
}
```

There must be a better way.

One possible direction is to make it convenient to use `CaptureEvent` directly. At the moment it is hard to replace a `CaptureException` call in user code with a `CaptureEvent` call because the former depends on unexported code.

Reference: https://stackoverflow.com/a/59898862/963287

**Comment by YashishDua** [comment]
@rhcarvalho Is this issue open to pick up? I have contributed in the early days of this repo.

**Comment by rhcarvalho** [comment]
@YashishDua before we can accept a PR for this there's some design to be done. We need some experimentation to find a better API that fits with the [Unified API](https://docs.sentry.io/development/sdk-dev/unified-api/) principles, if we are to find anything. Ideas welcome.

**Comment by YashishDua** [comment]
@rhcarvalho Alright! Would get back here soon.

**Comment by YashishDua** [comment]
@rhcarvalho I was going through the code base and Unified API principles and thinking if we can have an optional `level` argument for the `CaptureException` method. What do you think?

We can also look into allowing to add level types to EventHint maybe?

**Comment by rhcarvalho** [comment]
There are no "optional arguments" in Go. The closest to that are [_Variadic Functions_](https://golang.org/ref/spec#Function_types).

If we make a change like this: 

```diff
-func CaptureException(exception error) *EventID
+func CaptureException(exception error, level ...Level) *EventID
```

Then all these become possible:

```go
sentry.CaptureException(err)
sentry.CaptureException(err, LevelWarning)
sentry.CaptureException(err, LevelFatal, LevelWarning, LevelDebug)
```

An event can only have one level, now the last line above is confusing and unclear what level would be reported. I don't think this is a good direction to pursue.

**Comment by rhcarvalho** [comment]
> We can also look into allowing to add level types to EventHint maybe?

`sentry.CaptureException` and `Hub.CaptureException` currently don't take an `EventHint` and, even if it did, `EventHint` doesn't carry a `Level` :-(


As far as I know, nobody really complained about how to set levels as it is today. I opened the issue because I believe we can do better.

@YashishDua my suggestion would be to imagine examples of how one would write code to achieve the same as in the `foo` function from my original comment.

**Comment by rhcarvalho** [comment]
While writing the previous comment, I had a new idea: `Level` could implement `Capture*` methods, such this would work:

```go
func foo() {
    sentry.LevelWarning.CaptureException(errors.New("example warning"))
    sentry.LevelFatal.CaptureException(errors.New("example error"))
}
```

The ugly bit is "Exception" ... #161

**Comment by rhcarvalho** [comment]
And what I alluded to in the original issue comment:


```go
package sentry

func Capture(e Event) (id string) {
    // ...
}
```

```go
func foo() {
    sentry.Capture(Event{
        Level: LevelWarning,
        Error: errors.New("example warning"),
    })
    sentry.Capture(Event{
        Level: LevelFatal,
        Error: errors.New("example error"),
    })
}
```

**Comment by YashishDua** [comment]
@rhcarvalho 

> There are no "optional arguments" in Go. The closest to that are [_Variadic Functions_](https://golang.org/ref/spec#Function_types).
> 
> If we make a change like this:
> 
> ```diff
> -func CaptureException(exception error) *EventID
> +func CaptureException(exception error, level ...Level) *EventID
> ```
> 
> Then all these become possible:
> 
> ```go
> sentry.CaptureException(err)
> sentry.CaptureException(err, LevelWarning)
> sentry.CaptureException(err, LevelFatal, LevelWarning, LevelDebug)
> ```
> 
> An event can only have one level, now the last line above is confusing and unclear what level would be reported. I don't think this is a good direction to pursue.

I mean CaptureException would take one more argument apart from exception/error which is a  single value (as your above-mentioned way also takes a level per event). And if the argument value is not present, we can go with an error by default.

**Comment by rhcarvalho** [comment]
@YashishDua I'm very sorry I didn't follow, could you explain with an example code?

**Comment by YashishDua** [comment]
```go
type exceptionLevel struct {
    level Level
}

func createException(exception error, level *exceptionLevel) {}
```

Something of this sort, with better names.

**Comment by rhcarvalho** [comment]
@YashishDua ah, thanks. I think I got it.

Usage would be like

```go
sentry.CaptureException(err, nil)
```

```go
sentry.CaptureException(err, &sentry. LevelFatal)
```

And `CaptureException` would have to take the `nil` and transform it into a default value, `sentry.LevelError`.

**Comment by mariaefi29** [comment]
Dear developers, 

Do you have ETA for this feature? Would be very helpful in our application :)

**Comment by rhcarvalho** [comment]
Hi @mariaefi29! Unfortunately we cannot provide an ETA on this.

Would you be so kind and share how this would affect your application?

Note that it is possible to [set different levels on events](https://github.com/getsentry/sentry-go/issues/151#issue-556334376). I created this issue fundamentally to investigate if there is a better, more ergonomic if you will, API.

**Comment by mariaefi29** [comment]
Hi @rhcarvalho!

It is not so urgent of course) The only problem that I have is that I don't like how my code looks :) However, thank you for your examples, they've helped! 

I have a condition on status code that I take from http error from echo, so now it looks like:

```
	if err.(*echo.HTTPError).Code == http.StatusNotFound {
		sentry.WithScope(func(scope *sentry.Scope) {
			scope.SetLevel(sentry.LevelWarning)
			sentry.CaptureException(err)
		})
	} else {
		sentry.CaptureException(err)
	}
```
and it would be much better without a scope at all, cause I have a global scope set up in a middleware! 

For instance, something is like this is much better I guess :)
```
	if err.(*echo.HTTPError).Code == http.StatusNotFound {
			sentry.CaptureException(err, sentry.LevelWarning)
		})
	} else {
		sentry.CaptureException(err, sentry.LevelError)
	}
```

**Comment by rhcarvalho** [comment]
@mariaefi29, I see one alternative you can try: postpone setting the level in the `BeforeSend` function.

Your handler code would get simpler:

```go
	if err != nil {
		sentry.CaptureException(err)
	}
```

And when you initialize the SDK in `func main`, you can set `BeforeSend` to be something like:

```go
func beforeSend(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
	err := hint.OriginalException
	if err, ok := err.(*echo.HTTPError); ok && err.Code == http.StatusNotFound {
		event.Level = sentry.LevelWarning
	}
	return event
}

func main() {
	err := sentry.Init(sentry.ClientOptions{
		BeforeSend: beforeSend,
	})
	// ...
}
```

**Advantages**

- You only need to write `beforeSend` once, instead of writing the same logic multiple times in all request handlers.
- You can fine tune the events even further in `beforeSend`, endless possibilities (please don't over do it :smile:).
- No need to care about scopes and hubs.

**Disadvantages**

- Putting a lot of logic in `beforeSend` may make it confusing. You can mitigate this by writing several small functions that take care of one single aspect (e.g. making 404 errors into warnings), and then compose them in a function that calls all the others.
- Moving logic to `beforeSend` may make reading the code more "[magical](http://catb.org/jargon/html/M/magic.html)", in the sense that someone reading the request handler code has no idea that some further processing may happen before the event is sent (requires knowledge on how the SDK works to go look for `beforeSend`).

---

**Extra note**

Note that unless your program gives some guarantee that your error is always a `*echo.HTTPError`, the type assertion `err.(*echo.HTTPError).Code` may panic.

In the `beforeSend` implementation I shared above, we have `err, ok := err.(*echo.HTTPError)`, because at that point we certainly do not know what type the error is.

[The Go Programming Language Specification has the details on Type Assertions](https://golang.org/ref/spec#Type_assertions).

**Comment by mariaefi29** [comment]
Dear @rhcarvalho!

Thank you so much for your advice and wonderful examples!

To be honest, we are not comfortable having those custom types like api error codes from internal packages (like http and grpc servers) out in main.go file like that. That is why I rejected that idea in the first place when I started setting up levels on events.

Previously I had only fingerprints setup depending on error types in BeforeFunc in main.go file, and adding levels there seemed too much as well. However, I also wasn't fond of fingerprints in there (for the same reasons), but that was what sentry docs were saying to do. Yesterday I have found that you could set up fingerprints on scope as well in our internal packages! That was perfect! As a result, we have removed BeforeFunc out of main.go completely and it looks much more cleaner now. All the logic on custom types are in the same packages where those types were declared and used.

In addition, I don't write this:

```
if err != nil {
		sentry.CaptureException(err)
	}
``` 
in every handler, because we have set up the error logging middleware in echo server and notification interceptor in gRPC server) so really I write it once in every package in middleware) 

Thank you very much for your help! It gave me some ideas that led to the right code style for us! Very grateful!
